phidgets_steppers/src/steppers_ros_i.cpp

```diff
@@ -0,0 +1,381 @@
+             "set_velocity_limit%02d", (int)channel);
+    velocity_limit_subscription_ =
+        node->create_subscription<std_msgs::msg::Float64>(
+            velocity_limit_topicname, rclcpp::QoS(1),
+            std::bind(&StepperSetter::setVelocityLimitCallback, this,
+                      std::placeholders::_1));
+
+    char acceleration_topicname[] = "set_acceleration00";
+    snprintf(acceleration_topicname, sizeof(acceleration_topicname),
+             "set_acceleration%02d", (int)channel);
+    acceleration_subscription_ =
+        node->create_subscription<std_msgs::msg::Float64>(
+            acceleration_topicname, rclcpp::QoS(1),
+            std::bind(&StepperSetter::setAccelerationCallback, this,
+                      std::placeholders::_1));
+
+    char current_limit_topicname[] = "set_current_limit00";
+    snprintf(current_limit_topicname, sizeof(current_limit_topicname),
+             "set_current_limit%02d", (int)channel);
+    current_limit_subscription_ =
+        node->create_subscription<std_msgs::msg::Float64>(
+            current_limit_topicname, rclcpp::QoS(1),
+            std::bind(&StepperSetter::setCurrentLimitCallback, this,
+                      std::placeholders::_1));
+
+    char holding_current_limit_topicname[] = "set_holding_current_limit00";
+    snprintf(holding_current_limit_topicname,
+             sizeof(holding_current_limit_topicname),
+             "set_holding_current_limit%02d", (int)channel);
+    holding_current_limit_subscription_ =
+        node->create_subscription<std_msgs::msg::Float64>(
+            holding_current_limit_topicname, rclcpp::QoS(1),
+            std::bind(&StepperSetter::setHoldingCurrentLimitCallback, this,
+                      std::placeholders::_1));
+
+    char position_offset_topicname[] = "add_position_offset00";
+    snprintf(position_offset_topicname, sizeof(position_offset_topicname),
+             "add_position_offset%02d", (int)channel);
+    position_offset_subscription_ =
+        node->create_subscription<std_msgs::msg::Float64>(
+            position_offset_topicname, rclcpp::QoS(1),
+            std::bind(&StepperSetter::addPositionOffsetCallback, this,
+                      std::placeholders::_1));
+
+    char zero_position_topicname[] = "zero_position00";
+    snprintf(zero_position_topicname, sizeof(zero_position_topicname),
+             "zero_position%02d", (int)channel);
+    zero_position_subscription_ =
+        node->create_subscription<std_msgs::msg::Empty>(
+            zero_position_topicname, rclcpp::QoS(1),
+            std::bind(&StepperSetter::zeroPositionCallback, this,
+                      std::placeholders::_1));
+}
+
+void StepperSetter::setEnabledCallback(const std_msgs::msg::Bool::SharedPtr msg)
+{
+    try
+    {
+        steppers_->setEngaged(channel_, msg->data);
+    } catch (const phidgets::Phidget22Error& err)
+    {
+        // If the data was wrong, the lower layers will throw an exception; just
```

Comment:
Good point! I will switch these to services. Do you think it is better to return a boolean status or to return something like the set value? That might depend on how out-of-range values are handled. If a user tries to set a value that is outside the range, do you think it would be better to return an error or just coerce the value to be within the range and then return that value so the user knows what it is? I would also like to be able to ask what all of the stepper settings are. That could be a separate service or alternatively whenever one value is set, all of the settings could be returned every time.

Reply:
I think a boolean return is fine.  Also, I much prefer not to coerce values; instead, just try to do what the user asked and fail if it is out-of-range.  This could be made somewhat nicer by providing a way for the user to easily query what the range is, but it's probably not that important so I'd leave it out for now.